File: backend/methods/solver_params_extractor.py
"""
solver_params_extractor.py  –  Heuristic extraction of numerical solver inputs
from natural-language past exam questions.

Each extractor returns a dict that matches the solver's expected URL params so
the frontend can pre-fill the solver form.  Unrecognised questions return an
empty dict and the frontend falls back to sensible defaults.
"""

import re
from typing import Any

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _find_numbers(text: str) -> list[float]:
    """Return all numbers (incl. negative) found in text."""
    return [float(m) for m in re.findall(r"-?\d+(?:\.\d+)?(?:e[+-]?\d+)?", text)]


def _strip_latex(text: str) -> str:
    """Remove common LaTeX markup so regex can find the math."""
    text = re.sub(r"\\frac\{([^}]+)\}\{([^}]+)\}", r"(\1)/(\2)", text)
    text = re.sub(r"\\[a-zA-Z]+\*?", " ", text)
    text = re.sub(r"[{}]", " ", text)
    return text


def _latex_func_to_python(expr: str) -> str:
    """Best-effort conversion of a LaTeX-flavoured expression to Python syntax."""
    expr = re.sub(r"\\frac\{([^}]+)\}\{([^}]+)\}", r"(\1)/(\2)", expr)
    expr = re.sub(r"\\sqrt\{([^}]+)\}", r"(\1)**0.5", expr)
    expr = re.sub(r"\\left|\\right|\\,|\\;|\\!", "", expr)
    expr = re.sub(r"\^([0-9]+)", r"**\1", expr)        # x^3 → x**3
    expr = re.sub(r"\^(\{[^}]+\})", lambda m: f"**({m.group(1)[1:-1]})", expr)
    expr = re.sub(r"\\ln", "log", expr)
    expr = re.sub(r"\\log", "log10", expr)
    expr = re.sub(r"\\sin", "sin", expr)
    expr = re.sub(r"\\cos", "cos", expr)
    expr = re.sub(r"\\tan", "tan", expr)
    expr = re.sub(r"\\exp", "exp", expr)
    expr = re.sub(r"\\e\b", "exp(1)", expr)
    expr = re.sub(r"[{}]", "", expr)
    expr = re.sub(r"\s+", "", expr)
    return expr.strip()


def _extract_func_from_equation(question: str) -> str | None:
    """
    Try to extract f(x) = 0 style equations.
    e.g. "root of x^3 - x - 1 = 0" → "x**3 - x - 1"
    """
    patterns = [
        r"\$f\(x\)\s*=\s*([^$]+?)\$",                    # $f(x) = expr$
        r"f\(x\)\s*=\s*([^$,.\n]+)",                     # f(x) = expr
        r"equation\s+([^$]+?)\s*=\s*0",                  # equation ... = 0
        r"\$([^$]+?)\s*=\s*0\$",                         # $expr = 0$
        r"root of (?:the equation\s*)?(?:\$)?(.+?)\s*=\s*0",
        r"solve\s+(?:\$)?(.+?)\s*=\s*0",
    ]
    for pat in patterns:
        m = re.search(pat, question, re.IGNORECASE)
        if m:
            raw = m.group(1).strip().strip("$")
            if raw == "0" or raw.lower() == "y" or raw.lower() == "f(x)":
                continue
            py = _latex_func_to_python(raw)
            # Make sure we didn't just capture English text like "0 is"
            if py and len(py) >= 1 and not re.search(r'\bis\b|\bthe\b|\bof\b', py.lower()):
                return py
    return None


def _extract_ode_func(question: str) -> str | None:
    """
    Extract RHS of dy/dx = f(x,y) or y' = f.
    """
    patterns = [
        r"\\frac\{dy\}\{dx\}\s*=\s*(.+?)(?:\s*,|\s*\$|\s*with|\s*and|\s*$)",
        r"y'\s*=\s*(.+?)(?:\s*,|\s*\$|\s*with|\s*\.|\s*$)",
        r"y'\s*=\s*\$(.+?)\$",
    ]
    for pat in patterns:
        m = re.search(pat, question, re.IGNORECASE)
        if m:
            raw = m.group(1).strip().strip("$").strip(",").strip()
            py = _latex_func_to_python(raw)
            if py:
                return py
    return None


def _extract_integral_func(question: str) -> str | None:
    """
    Extract integrand from ∫_a^b f(x) dx style expressions.
    """
    # LaTeX: \int_0^1 f(x) \, dx
    m = re.search(r"\\int[_^{}\d\w]+\s+(.+?)\s*(?:\\,\s*)?d[xyzt]", question)
    if m:
        raw = m.group(1).strip().strip("$")
        return _latex_func_to_python(raw)

    # Plain text: integral of f(x) from
    m = re.search(r"integral of (.+?) (?:from|over|between)", question, re.IGNORECASE)
    if m:
        return _latex_func_to_python(m.group(1))

    return None


def _extract_bounds(question: str) -> tuple[float | None, float | None]:
    """Extract interval [a, b] from text."""
    stripped = _strip_latex(question)

    # "in [a, b]" or "in (a, b)" or "interval (a, b)"
    m = re.search(r"(?:in|interval|between)\s*[\[(](-?\d+\.?\d*)\s*,\s*(-?\d+\.?\d*)[\])]", stripped, re.IGNORECASE)
    if m:
        return float(m.group(1)), float(m.group(2))

    # "initial approximations a and b"
    m = re.search(r"approximations?\s+(-?\d+\.?\d*)\s+and\s+(-?\d+\.?\d*)", stripped, re.IGNORECASE)
    if m:
        return float(m.group(1)), float(m.group(2))

    # \int_a^b – bounds from LaTeX integral
    m = re.search(r"\\int_\{?(-?\d+\.?\d*)\}?\^\{?(-?\d+\.?\d*)\}?", question)
    if m:
        return float(m.group(1)), float(m.group(2))

    # from 0 to 1 style
    m = re.search(r"from\s+(-?\d+\.?\d*)\s+to\s+(-?\d+\.?\d*)", stripped, re.IGNORECASE)
    if m:
        return float(m.group(1)), float(m.group(2))

    return None, None


def _extract_x0_y0(question: str) -> tuple[float | None, float | None]:
    """Extract initial condition x0, y0."""
    stripped = _strip_latex(question)

    # y(x0) = y0 pattern
    m = re.search(r"y\s*\((-?\d+\.?\d*)\)\s*=\s*(-?\d+\.?\d*)", stripped)
    if m:
        return float(m.group(1)), float(m.group(2))

    # "given that x0 = N" or "x_0 = N"
    m = re.search(r"x[_0]*\s*=\s*(-?\d+\.?\d*)", stripped, re.IGNORECASE)
    x0 = float(m.group(1)) if m else None

    m2 = re.search(r"y[_0]*\s*=\s*(-?\d+\.?\d*)", stripped, re.IGNORECASE)
    y0 = float(m2.group(1)) if m2 else None

    return x0, y0


def _extract_step_size(question: str) -> float | None:
    stripped = _strip_latex(question)
    m = re.search(r"(?:step\s*size|h)\s*=\s*(-?\d+\.?\d*(?:/\d+)?)", stripped, re.IGNORECASE)
    if m:
        val = m.group(1)
        if '/' in val:
            parts = val.split('/')
            return float(parts[0]) / float(parts[1])
        return float(val)
    return None


def _extract_n(question: str) -> int | None:
    stripped = _strip_latex(question)
    m = re.search(r"n\s*=\s*(\d+)", stripped, re.IGNORECASE)
    if m:
        return int(m.group(1))
    # step size → derive n from bounds if possible
    return None


# ---------------------------------------------------------------------------
# Per-method extractors
# ---------------------------------------------------------------------------

def _params_bisection(question: str) -> dict[str, Any]:
    func = _extract_func_from_equation(question)
    a, b = _extract_bounds(question)
    return {
        "method": "bisection",
        "func_str": func or "x**3 - x - 1",
        "a": a if a is not None else 1.0,
        "b": b if b is not None else 2.0,
        "tol": 1e-4,
        "max_iter": 50,
    }


def _params_newton_raphson(question: str) -> dict[str, Any]:
    func = _extract_func_from_equation(question)
    x0, _ = _extract_x0_y0(question)
    # Try "x0 = N" style
    if x0 is None:
        stripped = _strip_latex(question)
        m = re.search(r"x_?0\s*=\s*(-?\d+\.?\d*)", stripped, re.IGNORECASE)
        if m:
            x0 = float(m.group(1))
    return {
        "method": "newton-raphson",
        "func_str": func or "x**3 - x - 1",
        "deriv_str": "",    # user must fill derivative
        "x0": x0 if x0 is not None else 1.5,
        "tol": 1e-6,
        "max_iter": 50,
    }


def _params_secant(question: str) -> dict[str, Any]:
    func = _extract_func_from_equation(question)
    # secant needs x0, x1 – try to grab two initial guesses
    stripped = _strip_latex(question)
    nums = _find_numbers(stripped)
    x0, x1 = (nums[0], nums[1]) if len(nums) >= 2 else (1.0, 2.0)
    return {
        "method": "secant",
        "func_str": func or "x**3 - x - 1",
        "x0": x0,
        "x1": x1,
        "tol": 1e-6,
        "max_iter": 50,
    }


def _params_false_position(question: str) -> dict[str, Any]:
    func = _extract_func_from_equation(question)
    a, b = _extract_bounds(question)
    return {
        "method": "false-position",
        "func_str": func or "x**3 - x - 1",
        "a": a if a is not None else 1.0,
        "b": b if b is not None else 2.0,
        "tol": 1e-4,
        "max_iter": 50,
    }


def _params_euler(question: str) -> dict[str, Any]:
    func = _extract_ode_func(question)
    x0, y0 = _extract_x0_y0(question)
    h = _extract_step_size(question)
    return {
        "method": "euler",
        "func_str": func or "x + y",
        "x0": x0 if x0 is not None else 0.0,
        "y0": y0 if y0 is not None else 1.0,
        "h": h if h is not None else 0.1,
        "steps_count": 5,
    }


def _params_modified_euler(question: str) -> dict[str, Any]:
    p = _params_euler(question)
    p["method"] = "modified-euler"
    return p


def _params_rk4(question: str) -> dict[str, Any]:
    func = _extract_ode_func(question)
    x0, y0 = _extract_x0_y0(question)
    h = _extract_step_size(question)
    return {
        "method": "rk4",
        "func_str": func or "x + y",
        "x0": x0 if x0 is not None else 0.0,
        "y0": y0 if y0 is not None else 1.0,
        "h": h if h is not None else 0.1,
        "steps_count": 5,
    }


def _params_trapezoidal(question: str) -> dict[str, Any]:
    func = _extract_integral_func(question)
    a, b = _extract_bounds(question)
    h = _extract_step_size(question)
    n = _extract_n(question)
    # If only h is known, derive n from bounds
    if n is None and h and a is not None and b is not None:
        n = max(1, round((b - a) / h))
    return {
        "method": "trapezoidal",
        "func_str": func or "1/(1+x)",
        "a": a if a is not None else 0.0,
        "b": b if b is not None else 1.0,
        "n": n if n is not None else 4,
    }


def _params_simpson_13(question: str) -> dict[str, Any]:
    p = _params_trapezoidal(question)
    p["method"] = "simpson-13"
    return p


def _params_simpson_38(question: str) -> dict[str, Any]:
    p = _params_trapezoidal(question)
    p["method"] = "simpson-38"
    if p["n"] and p["n"] % 3 != 0:
        p["n"] = max(3, (p["n"] // 3) * 3)
    return p


def _params_forward_difference(question: str) -> dict[str, Any]:
    """Interpolation — no automated solver yet, return empty."""
    return {"method": "forward-difference"}


# ---------------------------------------------------------------------------
# Dispatch table
# ---------------------------------------------------------------------------

_EXTRACTORS = {
    "bisection":          _params_bisection,
    "false-position":     _params_false_position,
    "newton-raphson":     _params_newton_raphson,
    "secant":             _params_secant,
    "euler":              _params_euler,
    "modified-euler":     _params_modified_euler,
    "rk4":                _params_rk4,
    "trapezoidal":        _params_trapezoidal,
    "simpson-13":         _params_simpson_13,
    "simpson-38":         _params_simpson_38,
    "forward-difference": _params_forward_difference,
    "lu-decomposition":   lambda _: {"method": "lu-decomposition"},
    "gauss-seidel":       lambda _: {"method": "gauss-seidel"},
    "gauss-jacobi":       lambda _: {"method": "gauss-jacobi"},
    "thomas-algorithm":   lambda _: {"method": "thomas-algorithm"},
}


def extract_solver_params(question_text: str, solver_method: str) -> dict[str, Any]:
    """
    Main entry point.  Returns a dict of pre-filled solver params for the given
    solver_method, extracted heuristically from question_text.
    """
    extractor = _EXTRACTORS.get(solver_method)
    if extractor is None:
        return {}
    try:
        return extractor(question_text)
    except Exception:
        # Never crash the API; just return empty and let the solver use defaults
        return {"method": solver_method}

File: worker/src/solverParamsExtractor.ts
// Port of backend/methods/solver_params_extractor.py
// Heuristic extraction of numerical solver inputs from natural-language exam questions.

type Params = Record<string, unknown>;

function stripChars(s: string, chars: string): string {
  let start = 0;
  let end = s.length;
  while (start < end && chars.includes(s[start])) start++;
  while (end > start && chars.includes(s[end - 1])) end--;
  return s.slice(start, end);
}

function findNumbers(text: string): number[] {
  const matches = text.match(/-?\d+(?:\.\d+)?(?:e[+-]?\d+)?/g);
  return matches ? matches.map(Number) : [];
}

function stripLatex(text: string): string {
  text = text.replace(/\\frac\{([^}]+)\}\{([^}]+)\}/g, "($1)/($2)");
  text = text.replace(/\\[a-zA-Z]+\*?/g, " ");
  text = text.replace(/[{}]/g, " ");
  return text;
}

function latexFuncToPython(expr: string): string {
  expr = expr.replace(/\\frac\{([^}]+)\}\{([^}]+)\}/g, "($1)/($2)");
  expr = expr.replace(/\\sqrt\{([^}]+)\}/g, "($1)**0.5");
  expr = expr.replace(/\\left|\\right|\\,|\\;|\\!/g, "");
  expr = expr.replace(/\^([0-9]+)/g, "**$1");
  expr = expr.replace(/\^(\{[^}]+\})/g, (_m, g1: string) => `**(${g1.slice(1, -1)})`);
  expr = expr.replace(/\\ln/g, "log");
  expr = expr.replace(/\\log/g, "log10");
  expr = expr.replace(/\\sin/g, "sin");
  expr = expr.replace(/\\cos/g, "cos");
  expr = expr.replace(/\\tan/g, "tan");
  expr = expr.replace(/\\exp/g, "exp");
  expr = expr.replace(/\\e\b/g, "exp(1)");
  expr = expr.replace(/[{}]/g, "");
  expr = expr.replace(/\s+/g, "");
  return expr.trim();
}

function extractFuncFromEquation(question: string): string | null {
  const patterns: RegExp[] = [
    /\$f\(x\)\s*=\s*([^$]+?)\$/i,
    /f\(x\)\s*=\s*([^$,.\n]+)/i,
    /equation\s+([^$]+?)\s*=\s*0/i,
    /\$([^$]+?)\s*=\s*0\$/i,
    /root of (?:the equation\s*)?(?:\$)?(.+?)\s*=\s*0/i,
    /solve\s+(?:\$)?(.+?)\s*=\s*0/i,
  ];

  for (const pat of patterns) {
    const m = question.match(pat);
    if (m) {
      const raw = stripChars(m[1].trim(), "$");
      if (raw === "0" || raw.toLowerCase() === "y" || raw.toLowerCase() === "f(x)") continue;
      const py = latexFuncToPython(raw);
      if (py && py.length >= 1 && !/\bis\b|\bthe\b|\bof\b/.test(py.toLowerCase())) {
        return py;
      }
    }
  }
  return null;
}

function extractOdeFunc(question: string): string | null {
  const patterns: RegExp[] = [
    /\\frac\{dy\}\{dx\}\s*=\s*(.+?)(?:\s*,|\s*\$|\s*with|\s*and|\s*$)/i,
    /y'\s*=\s*(.+?)(?:\s*,|\s*\$|\s*with|\s*\.|\s*$)/i,
    /y'\s*=\s*\$(.+?)\$/i,
  ];

  for (const pat of patterns) {
    const m = question.match(pat);
    if (m) {
      const raw = stripChars(stripChars(m[1].trim(), "$").trim(), ",").trim();
      const py = latexFuncToPython(raw);
      if (py) return py;
    }
  }
  return null;
}

function extractIntegralFunc(question: string): string | null {
  let m = question.match(/\\int[_^{}\d\w]+\s+(.+?)\s*(?:\\,\s*)?d[xyzt]/);
  if (m) {
    const raw = stripChars(m[1].trim(), "$");
    return latexFuncToPython(raw);
  }

  m = question.match(/integral of (.+?) (?:from|over|between)/i);
  if (m) return latexFuncToPython(m[1]);

  return null;
}

function extractBounds(question: string): [number | null, number | null] {
  const stripped = stripLatex(question);

  let m = stripped.match(/(?:in|interval|between)\s*[\[(](-?\d+\.?\d*)\s*,\s*(-?\d+\.?\d*)[\])]/i);
  if (m) return [parseFloat(m[1]), parseFloat(m[2])];

  m = stripped.match(/approximations?\s+(-?\d+\.?\d*)\s+and\s+(-?\d+\.?\d*)/i);
  if (m) return [parseFloat(m[1]), parseFloat(m[2])];

  m = question.match(/\\int_\{?(-?\d+\.?\d*)\}?\^\{?(-?\d+\.?\d*)\}?/);
  if (m) return [parseFloat(m[1]), parseFloat(m[2])];

  m = stripped.match(/from\s+(-?\d+\.?\d*)\s+to\s+(-?\d+\.?\d*)/i);
  if (m) return [parseFloat(m[1]), parseFloat(m[2])];

  return [null, null];
}

function extractX0Y0(question: string): [number | null, number | null] {
  const stripped = stripLatex(question);

  let m = stripped.match(/y\s*\((-?\d+\.?\d*)\)\s*=\s*(-?\d+\.?\d*)/);
  if (m) return [parseFloat(m[1]), parseFloat(m[2])];

  m = stripped.match(/x[_0]*\s*=\s*(-?\d+\.?\d*)/i);
  const x0 = m ? parseFloat(m[1]) : null;

  const m2 = stripped.match(/y[_0]*\s*=\s*(-?\d+\.?\d*)/i);
  const y0 = m2 ? parseFloat(m2[1]) : null;

  return [x0, y0];
}

function extractStepSize(question: string): number | null {
  const stripped = stripLatex(question);
  const m = stripped.match(/(?:step\s*size|h)\s*=\s*(-?\d+\.?\d*(?:\/\d+)?)/i);
  if (m) {
    const val = m[1];
    if (val.includes("/")) {
      const [num, den] = val.split("/");
      return parseFloat(num) / parseFloat(den);
    }
    return parseFloat(val);
  }
  return null;
}

function extractN(question: string): number | null {
  const stripped = stripLatex(question);
  const m = stripped.match(/n\s*=\s*(\d+)/i);
  if (m) return parseInt(m[1], 10);
  return null;
}

// ---------------------------------------------------------------------------
// Per-method extractors
// ---------------------------------------------------------------------------

function paramsBisection(question: string): Params {
  const func = extractFuncFromEquation(question);
  const [a, b] = extractBounds(question);
  return {
    method: "bisection",
    func_str: func ?? "x**3 - x - 1",
    a: a ?? 1.0,
    b: b ?? 2.0,
    tol: 1e-4,
    max_iter: 50,
  };
}

function paramsNewtonRaphson(question: string): Params {
  const func = extractFuncFromEquation(question);
  let [x0] = extractX0Y0(question);
  if (x0 === null) {
    const stripped = stripLatex(question);
    const m = stripped.match(/x_?0\s*=\s*(-?\d+\.?\d*)/i);
    if (m) x0 = parseFloat(m[1]);
  }
  return {
    method: "newton-raphson",
    func_str: func ?? "x**3 - x - 1",
    deriv_str: "",
    x0: x0 ?? 1.5,
    tol: 1e-6,
    max_iter: 50,
  };
}

function paramsSecant(question: string): Params {
  const func = extractFuncFromEquation(question);
  const stripped = stripLatex(question);
  const nums = findNumbers(stripped);
  const [x0, x1] = nums.length >= 2 ? [nums[0], nums[1]] : [1.0, 2.0];
  return {
    method: "secant",
    func_str: func ?? "x**3 - x - 1",
    x0,
    x1,
    tol: 1e-6,
    max_iter: 50,
  };
}

function paramsFalsePosition(question: string): Params {
  const func = extractFuncFromEquation(question);
  const [a, b] = extractBounds(question);
  return {
    method: "false-position",
    func_str: func ?? "x**3 - x - 1",
    a: a ?? 1.0,
    b: b ?? 2.0,
    tol: 1e-4,
    max_iter: 50,
  };
}

function paramsEuler(question: string): Params {
  const func = extractOdeFunc(question);
  const [x0, y0] = extractX0Y0(question);
  const h = extractStepSize(question);
  return {
    method: "euler",
    func_str: func ?? "x + y",
    x0: x0 ?? 0.0,
    y0: y0 ?? 1.0,
    h: h ?? 0.1,
    steps_count: 5,
  };
}

function paramsModifiedEuler(question: string): Params {
  const p = paramsEuler(question);
  p.method = "modified-euler";
  return p;
}

function paramsRk4(question: string): Params {
  const func = extractOdeFunc(question);
  const [x0, y0] = extractX0Y0(question);
  const h = extractStepSize(question);
  return {
    method: "rk4",
    func_str: func ?? "x + y",
    x0: x0 ?? 0.0,
    y0: y0 ?? 1.0,
    h: h ?? 0.1,
    steps_count: 5,
  };
}

function paramsTrapezoidal(question: string): Params {
  const func = extractIntegralFunc(question);
  const [a, b] = extractBounds(question);
  const h = extractStepSize(question);
  let n = extractN(question);
  if (n === null && h && a !== null && b !== null) {
    n = Math.max(1, Math.round((b - a) / h));
  }
  return {
    method: "trapezoidal",
    func_str: func ?? "1/(1+x)",
    a: a ?? 0.0,
    b: b ?? 1.0,
    n: n ?? 4,
  };
}

function paramsSimpson13(question: string): Params {
  const p = paramsTrapezoidal(question);
  p.method = "simpson-13";
  return p;
}

function paramsSimpson38(question: string): Params {
  const p = paramsTrapezoidal(question);
  p.method = "simpson-38";
  const n = p.n as number;
  if (n && n % 3 !== 0) {
    p.n = Math.max(3, Math.floor(n / 3) * 3);
  }
  return p;
}

function paramsForwardDifference(_question: string): Params {
  return { method: "forward-difference" };
}

// ---------------------------------------------------------------------------
// Dispatch table
// ---------------------------------------------------------------------------

const EXTRACTORS: Record<string, (question: string) => Params> = {
  bisection: paramsBisection,
  "false-position": paramsFalsePosition,
  "newton-raphson": paramsNewtonRaphson,
  secant: paramsSecant,
  euler: paramsEuler,
  "modified-euler": paramsModifiedEuler,
  rk4: paramsRk4,
  trapezoidal: paramsTrapezoidal,
  "simpson-13": paramsSimpson13,
  "simpson-38": paramsSimpson38,
  "forward-difference": paramsForwardDifference,
  "lu-decomposition": () => ({ method: "lu-decomposition" }),
  "gauss-seidel": () => ({ method: "gauss-seidel" }),
  "gauss-jacobi": () => ({ method: "gauss-jacobi" }),
  "thomas-algorithm": () => ({ method: "thomas-algorithm" }),
};

export function extractSolverParams(questionText: string, solverMethod: string): Params {
  const extractor = EXTRACTORS[solverMethod];
  if (!extractor) return {};
  try {
    return extractor(questionText);
  } catch {
    return { method: solverMethod };
  }
}
